test/FsAutoComplete.Tests.Lsp/InlayHintTests.fs

```diff
@@ -550,3 +547,1139 @@ let tests state =
+
+        for (actual, (expected, textAfterEdits)) in Seq.zip hints expected do
+          do! validateHint doc expected textAfterEdits text actual
+      }
+
+      do! checkInRange server text range validateHints
+    }
+
+  let private fromCursor: Position = { Line = -1; Character = -1 }
+
+  let private mkBasicHint
+    (kind: InlayHintKind)
+    (pos: Position)
+    (label: string)
+    : InlayHint
+    =
+    {
+      Kind = Some kind
+      Position = pos
+      Label = InlayHintLabel.String label
+      TextEdits = None
+      Tooltip = None
+      PaddingLeft = match kind with | InlayHintKind.Type -> Some true | _ -> None
+      PaddingRight = match kind with | InlayHintKind.Parameter -> Some true | _ -> None
+      Data = None
+    }
+  let paramHint
+    (paramName: string)
+    =
+    let label = $"{paramName} ="
+    let hint = mkBasicHint InlayHintKind.Parameter fromCursor label
+    (hint, None)
+  let typeHint
+    (typeName: string)
+    (expectedAfterEdits: string)
+    =
+    let label = $": {typeName}"
+    let hint = mkBasicHint InlayHintKind.Type fromCursor label
+    let expectedAfterEdits =
+      expectedAfterEdits
+      |> Text.trimTripleQuotation
+    (hint, Some expectedAfterEdits)
+
+open LspInlayHints
+let private paramHintTests state =
+  serverTestList "param hints" state defaultConfigDto None (fun server -> [
+    testCaseAsync "can show param hint" <|
+      checkAllInMarkedRange server
+        """
+        let f beta = ()
+        $|f $042$|
+        """
+        [
+          paramHint "beta"
+        ]
+  ])
+let private typeHintTests state =
+  serverTestList "type hints" state defaultConfigDto None (fun server -> [
+    testCaseAsync "can show type hint" <|
+      checkAllInMarkedRange server
+        """
+        $|let f beta$0 = beta + 1$|
+        """
+        [
+          typeHint "int"
+            """
+            let f (beta: int) = beta + 1
+            """
+        ]
+    testCaseAsync "can show type for generic actual type" <|
+      checkAllInMarkedRange server
+        """
+        open System.Collections.Generic
+        $|let list$0 = List<int>()$|
+        list.Add 2
+        """
+        [
+          typeHint "List<int>"
+            """
+            open System.Collections.Generic
+            let list: List<int> = List<int>()
+            list.Add 2
+            """
+        ]
+    ptestCaseAsync "can show type hint for nested inside generic actual type" <|
+      checkAllInMarkedRange server
+        """
+        open System.Collections.Immutable
+        $|let arr$0 = ImmutableArray.CreateBuilder()$|
+        arr.Add 2
+        """
+        [
+          //Currently: `ImmutableArray`1.Builder<int>`
+          typeHint "ImmutableArray<int>.Builder"
+            """
+            open System.Collections.Immutable
+            let arr: ImmutableArray<int>.Builder = ImmutableArray.CreateBuilder()
+            arr.Add 2
+            """
+        ]
```

Comment:
Is there a way to format the type correctly?  
``ImmutableArray`1.Builder<int>`` isn't valid.


Considering the tooltip in VS shows that as well, and nested types aren't possible in F# to create, I don't think there's something that handles type formatting correctly?

Reply:
I think you're right and there's a gap here in the type formatting from FCS for these types.